### PR TITLE
Distinguish between "rendered" and "unrendered" Tensors.

### PR DESCRIPTION
**Pull request by judah**
There are now three types of `Tensor`:

- `Tensor Value a`: rendered value
- `Tensor Ref a`: rendered reference
- `Tensor Build a` : unrendered value

The extra bookkeeping makes it easier to track (and enforce) which tensors are
rendered or not.  For examples where this has been confusing in the past, see

With this change, pure ops look similar to before, returning `Tensor Build`
instead of `Tensor Value`.  "Stateful" (monadic) ops are unchanged.  For
example:

    add :: OneOf [..] t => Tensor v'1 t -> Tensor v'2 t -> Tensor Build t
    assign :: (MonadBuild m, TensorType t)
           => Tensor Ref t -> Tensor v'2 t -> m (Tensor Ref t)

The `gradients` function now requires that the variables over which it's
differentiating are pre-rendered:

    gradients :: (..., Rendered v2) => Tensor v1 a -> [Tensor v2 a]
              -> m [Tensor Value a]

(`Rendered v2` means that `v2` is either a `Ref` or a `Value`.)

Additionally, the implementation of `gradients` now takes care to render every
intermediate value when performing the reverse accumulation.  I suspect this
fixes an exponential blowup for complicated expressions.

### CODE REVIEW

File: tensorflow-opgen/src/TensorFlow/OpGen.hs [hunk 1]
```diff
@@ -259,10 +262,9 @@ functionBody pOp = maybeLift <+> buildFunction <+> parens (hang 0 (stack buildOp
         [ "& opAttr" <+> renderQuotedTFName n <+> ".~" <+> renderHaskellName n
         | a <- inferredListSizeAttrs pOp, let n = attrName a
         ] ++
-        ["& op'options"]
-
-
-    tensorArgs = renderHaskellName . parsedArgName <$> parsedInputs pOp
+        ["& op'options & opInputs .~ op'inputs"]
```
Comment (fkm3) [inline tensorflow-opgen/src/TensorFlow/OpGen.hs]: Maybe use `opInputsVar` here.

File: tensorflow/src/TensorFlow/Build.hs [hunk 2]
```diff
@@ -317,12 +308,11 @@ renderPendingNode (PendingNode scope pendingName nodeDef)
...
--- | Render an 'Output' and return a string representation for the TensorFlow
+-- | Turn an 'Output' into a string representation for the TensorFlow
 -- foreign APIs.
-renderOutput :: Output -> Build Text
-renderOutput (Output (OutputIx i) o) = do
-    n <- getOrAddOp o
-    return $ unNodeName n <> Text.pack (":" ++ show i)
+encodeOutput :: Output -> Text
+encodeOutput (Output (OutputIx 0) n) = unNodeName n
+encodeOutput (Output (OutputIx i) n) = unNodeName n <> Text.pack (":" ++ show i)
```
Comment (judah) [inline tensorflow/src/TensorFlow/Build.hs]: Done.

File: tensorflow/src/TensorFlow/Nodes.hs [hunk 3]
```diff
@@ -113,13 +113,13 @@ instance (Fetchable (f t) a, Fetchable (ListOf f ts) (List as), i ~ Identity)
     getFetch (x :/ xs) = liftA2 (\y ys -> y /:/ ys) <$> getFetch x <*> getFetch xs
 
 instance Nodes (Tensor v a) where
-    getNodes t = Set.singleton <$> getOrAddOp (t ^. tensorOutput . outputOp)
+    getNodes (Tensor o) = Set.singleton . outputNodeName <$> toBuild o
 
-fetchTensorVector :: forall a v . TensorType a
+fetchTensorVector :: forall a v . (TensorType a)
```
Comment (judah) [inline tensorflow/src/TensorFlow/Nodes.hs]: I prefer to avoid the parentheses when they're unnecessary as in this case.

File: tensorflow-ops/src/TensorFlow/Ops.hs [hunk 4]
```diff
@@ -261,20 +263,23 @@ restoreFromName :: forall a m . (MonadBuild m, TensorType a)
...
 restore :: forall a m . (MonadBuild m, TensorType a)
         => ByteString    -- ^ File path.
         -> Tensor Ref a  -- ^ Tensor to restore.
         -> m ControlNode
-restore path x = do
-    name <- encodeUtf8 . unNodeName <$> build (renderNodeName x)
-    restoreFromName path name x
+restore path x = restoreFromName path name x
+  where
+    -- TODO: renderTensor?
```
Comment (judah) [inline tensorflow-ops/src/TensorFlow/Ops.hs]: Removed; it's obsolete.

File: tensorflow-ops/src/TensorFlow/Ops.hs [hunk 5]
```diff
@@ -368,3 +373,8 @@ reducedShape inputShape axes =
            axesMod]                               -- [1, 2]
          [inputShape32,                           -- [2, 3, 5, 7]
            CoreOps.fill axesShape 1]              -- [1, 1]
+
+-- OK: this doesn't work: passing integer literal into op makes it fail...sigh...
+-- options:
+-- - move back to explicit "expr"
+-- - use defaulting
```
Comment (judah) [inline tensorflow-ops/src/TensorFlow/Ops.hs]: Removed, sorry for missing these comments earlier.

File: tensorflow-nn/src/TensorFlow/NN.hs [hunk 6]
```diff
@@ -76,13 +78,11 @@ sigmoidCrossEntropyWithLogits
...
-        neg_abs_logits = select cond (-logits') logits'
+    let zeros = zerosLike logits
+        cond = logits `greaterEqual` zeros
+        relu_logits = select cond logits zeros
+        neg_abs_logits = select cond (neg logits) logits
     withNameScope "logistic_loss" $ do
-        left  <- render $ relu_logits - logits' * targets'
-        right <- render $ log (1 + exp neg_abs_logits)
+        let left = relu_logits - logits `mul` targets
+        let right = log (1 + exp neg_abs_logits)
```
Comment (judah) [inline tensorflow-nn/src/TensorFlow/NN.hs]: Ah, ok.  Reverted to the previous behavior.

File: tensorflow-ops/tests/EmbeddingOpsTest.hs [hunk 7]
```diff
@@ -48,16 +49,15 @@ testEmbeddingLookupHasRightShapeWithPartition =
...
     let idValues  = [0, 1 :: Int32]
-    let ids       = TF.constant (TF.Shape [1, 2]) idValues
-    let op        = embeddingLookup embedding ids
 
     (values, shape) <- TF.runSession $ do
-        vs <- op
+        embedding   <- mapM TF.render [ TF.constant embShape embedding1
+                        , TF.constant embShape embedding2
+                        ]
+        ids       <- TF.render $ TF.constant (TF.Shape [1, 2]) idValues
```
Comment (judah) [inline tensorflow-ops/tests/EmbeddingOpsTest.hs]: I changed `ids` to be unrendered (and updated the type of embeddingLookup to allow that).

`embedding` needs to be rendered since we colocate ops with it.

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 8]
```diff
@@ -86,21 +86,25 @@ import TensorFlow.Ops
...
+    , Rendered
+    , tensorNodeName
+    , renderedOutput
+    , renderValue
     )
 import TensorFlow.Types (Attribute, OneOf, TensorType, attrLens)
 import Proto.Tensorflow.Core.Framework.NodeDef
     (NodeDef, attr, input, op, name)
 
+-- TODO: cache *rendered* gradients.
```
Comment (judah) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Obsolete comment; removed.

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 9]
```diff
@@ -397,46 +404,49 @@ type GradientFunc a = NodeDef
...
-                    -> [Maybe (Tensor Value a)]
+                    -> [Maybe (Tensor Build a)]
                     -- ^ Gradient of y w.r.t. each input tensor.
 
 
 -- TODO(fmayle): Assert the type is correct.
 -- | Create a Tensor from an Output.
-toT :: Output -> Tensor Value a
-toT = Tensor ValueKind
+toT :: Output -> Tensor Build a
```
Comment (judah) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Yeah, I think that was pretty much the reason.

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 10]
```diff
@@ -510,22 +520,22 @@ opGrad "SoftmaxCrossEntropyWithLogits" _ [toT -> x, toT -> y] [dz, _] =
...
+    sy = shape (y :: Tensor Build a)
     (rx, ry) = broadcastGradientArgs sx sy
 
 opGrad "Div" _ [toT -> x, toT -> y] [dz] =
     -- TODO(fmayle): Handle complex numbers.
     -- TODO(gnezdo): Provide Fractional instance and use '/' instead of div.
-    [ Just $ reshape (sum (dz `CoreOps.div` y) rx) sx
-    , Just $ reshape (sum (dz * (negate x `CoreOps.div` (y * y))) ry) sy
+    [ Just $ reshape (sum (expr dz `CoreOps.div` y) rx) sx
+    , Just $ reshape (sum (expr dz `CoreOps.mul` ((negate (expr x) :: Tensor Build a) `CoreOps.div` (expr y * expr y :: Tensor Build a) :: Tensor Build a)) ry) sy
```
Comment (judah) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Done, and removed the unnecessary type signatures.

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 11]
```diff
@@ -639,18 +649,18 @@ opGrad "Select" _ [toT -> c, toT -> x, _] [dz] =
   where zeros = CoreOps.zerosLike x
 
 -- TODO(gnezdo): Unlike Python, no control dependency on dz.
-opGrad "Log" _ [toT -> x] [dz] = [ Just $ dz * CoreOps.inv x ]
+opGrad "Log" _ [toT -> x] [dz] = [ Just $ dz `CoreOps.mul` CoreOps.inv x ]
 -- TODO(gnezdo): Reuse the output instead of doing another exp,
 -- though, it is probably CSE'd away anyway.
-opGrad "Exp" _ [toT -> x] [dz] = [ Just $ dz * CoreOps.exp x ]
+opGrad "Exp" _ [toT -> x] [dz] = [ Just $ dz `CoreOps.mul` CoreOps.exp (expr x) ]
```
Comment (judah) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Done.

File: tensorflow-opgen/src/TensorFlow/OpGen.hs (hunk 1 above)
Comment (judah) [inline tensorflow-opgen/src/TensorFlow/OpGen.hs]: Done.

File: tensorflow/src/TensorFlow/Nodes.hs (hunk 3 above)
Comment (fkm3) [inline tensorflow/src/TensorFlow/Nodes.hs]: I was recommending to remove the extra parens.

Comment (judah) [inline tensorflow/src/TensorFlow/Nodes.hs]: Oh, sorry, for some reason I misread the diff.  I'll send a follow-up.
